File: app/src/main/java/ardjomand/leonardo/nutrimeal/users/UsersAdapter.java
package ardjomand.leonardo.nutrimeal.users;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import ardjomand.leonardo.nutrimeal.R;
import ardjomand.leonardo.nutrimeal.auth.User;
import ardjomand.leonardo.nutrimeal.meals.Meal;
import ardjomand.leonardo.nutrimeal.meals.MealsFragment.OnMealFragmentInteractionListener;

/**
 * {@link RecyclerView.Adapter} that can display a {@link Meal} and makes a call to the
 * specified {@link OnMealFragmentInteractionListener}.
 */
public class UsersAdapter extends RecyclerView.Adapter<UsersAdapter.ViewHolder> {

    private final OnMealAdapterInteractionListener mListener;
    private List<User> mItems;

    UsersAdapter(List<User> orders, OnMealAdapterInteractionListener listener, Context context) {
        mItems = orders;
        mListener = listener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_user, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final ViewHolder holder, int position) {
        holder.mItem = mItems.get(position);

        holder.mNameView.setText(holder.mItem.getName());

        holder.mView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (null != mListener) {
                    // Notify the active callbacks interface (the activity, if the
                    // fragment is attached to one) that an item has been selected.
                    mListener.onUserClicked(holder.mItem);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public void replaceData(List<User> orders) {
        mItems = orders;
        notifyDataSetChanged();
    }

    public void addData(User user) {
        mItems.add(user);
        notifyItemInserted(mItems.size() - 1);
    }

    public void clearData() {
        mItems.clear();
        notifyDataSetChanged();
    }

    public void updateData(User user) {
        int index = getIndexForKey(user.getKey());
        mItems.set(index, user);
        notifyItemChanged(index);
    }

    private int getIndexForKey(String key) {
        int index = 0;
        for (User user : mItems) {
            if (user.getKey().equals(key)) {
                return index;
            } else {
                index++;
            }
        }
        throw new IllegalArgumentException("Key not found");
    }

    public interface OnMealAdapterInteractionListener {
        void onUserClicked(User item);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        final View mView;
        final TextView mNameView;
        User mItem;

        ViewHolder(View view) {
            super(view);
            mView = view;
            mNameView = view.findViewById(R.id.user_name);
        }

        @Override
        public String toString() {
            return super.toString() + " '" + mNameView.getText() + "'";
        }
    }
}

File: app/src/main/java/ardjomand/leonardo/nutrimeal/users/UsersAdapter.kt
package ardjomand.leonardo.nutrimeal.users

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import ardjomand.leonardo.nutrimeal.R
import ardjomand.leonardo.nutrimeal.Utils
import ardjomand.leonardo.nutrimeal.data.pojos.User

class UsersAdapter internal constructor(private var mItems: MutableList<User>?, private val mListener: OnMealAdapterInteractionListener?, context: Context) : RecyclerView.Adapter<UsersAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_user, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val mItem = mItems!![position]

        holder.mNameView.text = mItem.name

        holder.mView.setOnClickListener {
            mListener?.onUserClicked(mItem)
        }
    }

    override fun getItemCount(): Int {
        return mItems!!.size
    }

    fun addData(user: User) {
        mItems!!.add(user)
        notifyItemInserted(mItems!!.size - 1)
    }

    fun clearData() {
        mItems!!.clear()
        notifyDataSetChanged()
    }

    fun updateData(user: User) {
        val index = Utils.getIndexForKey(user.key, mItems!!)
        mItems!![index] = user
        notifyItemChanged(index)
    }

    interface OnMealAdapterInteractionListener {
        fun onUserClicked(item: User)
    }

    inner class ViewHolder internal constructor(internal val mView: View) : RecyclerView.ViewHolder(mView) {
        internal val mNameView: TextView = mView.findViewById(R.id.user_name)

        override fun toString(): String {
            return super.toString() + " ' + ${mNameView.text} + '"
        }
    }
}
